### PR TITLE
Package Manager Testing 1:  Refactoring of PackageUploadBuilder

### PR DESCRIPTION
**Pull request by pboyer**
### Issue

This is the first pull request of several to improve testing coverage on the package manager.  

This is the most significant refactoring and and focuses on `PackageUploadBuilder` and indirectly on `PackageManagerClient`.  

The "todo" should make it clear what I've changed and the state of the PR.  Future PR's will focus less on refactoring and more on extending test coverage.  

### TODO for this PR

- [x] Separated `PackageUploadBuilder` out into `PackageDirectoryBuilder`.
- [x] Added mockable interfaces to make it possible to mock mutable operations like modifications to the file system or file compression.
- [x]  Complete tests for `PackageUploadBuilder` and `PackageManagerClient`  

### Reviewer

@ikeough

### CODE REVIEW

File: test/DynamoCoreTests/PackageManager/PackageManagerClientTests.cs [hunk 1]
```diff
@@ -33,7 +34,7 @@ public void SearchWithEmptyQueryReturnsAllPackagesInSortedOrder()
         [Test]
         public void ListAllReturnsAllPackages()
         {
-            var elements = ViewModel.Model.PackageManagerClient.ListAll();
+            var elements = this.CurrentDynamoModel.PackageManagerClient.ListAll();
             Assert.AreNotEqual(0, elements.Count());
             Console.WriteLine(elements.Count());
         }
```
Comment (pboyer) [inline test/DynamoCoreTests/PackageManager/PackageManagerClientTests.cs]: `PackageManagerClientTests` will be done in a separate PR.  It is now thoroughly mocked.

File: test/DynamoCoreTests/PackageManager/PackageUploadBuilderTests.cs [hunk 2]
```diff
@@ -1,71 +1,175 @@
...
+        public void RemapCustomNodeFilePaths_SucceedsForValidFiles()
         {
             Assert.Inconclusive("Finish me");
         }
 
         [Test]
-        public void RemapCustomNodeFilePathsFailsForInvalidFiles()
+        public void RemapCustomNodeFilePaths_FailsForInvalidFiles()
         {
             Assert.Inconclusive("Finish me");
```
Comment (pboyer) [inline test/DynamoCoreTests/PackageManager/PackageUploadBuilderTests.cs]: This is far from done.  This will be updated in subsequent commits to this PR.

File: src/DynamoCore/PackageManager/PackageUploadBuilder.cs [hunk 3]
```diff
@@ -1,205 +1,93 @@
...
-using Dynamo.Utilities;
+using System.Text;
 using Greg.Requests;
-using RestSharp.Serializers;
 
 namespace Dynamo.PackageManager
 {
-    static class PackageUploadBuilder
+    internal class PackageUploadBuilder
     {
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageUploadBuilder.cs]: This is the most substantially changed file in this PR.  Most of it's guts are moved to `PackageDirectoryBuilder`.

File: src/DynamoCore/PackageManager/PackageUploadBuilder.cs [hunk 4]
```diff
@@ -1,205 +1,93 @@
...
-            var engineMetadata = "";
+        private readonly PackageDirectoryBuilder builder;
+        private readonly IFileCompressor fileCompressor;
 
-            return new PackageUploadRequestBody(l.Name, l.VersionName, l.Description, l.Keywords, l.License, l.Contents, "dynamo",
-                                                         engineVersion, engineMetadata, l.Group, l.Dependencies, 
-                                                         l.SiteUrl, l.RepositoryUrl, l.ContainsBinaries, l.NodeLibraries.Select(x => x.FullName) ); 
-        } 
+        internal const string PackageEngineName = "dynamo";
+        internal const long MaximumPackageSize = 100 * 1024 * 1024;
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageUploadBuilder.cs]: These were previously scatted around the code. Now they are `const`.

File: src/DynamoCore/PackageManager/PackageUploadBuilder.cs [hunk 5]
```diff
@@ -1,205 +1,93 @@
...
+
+            return new PackageUploadRequestBody(l.Name, l.VersionName, l.Description, l.Keywords, l.License, l.Contents, PackageEngineName,
+                                                         engineVersion, engineMetadata, l.Group, l.Dependencies,
+                                                         l.SiteUrl, l.RepositoryUrl, l.ContainsBinaries, l.NodeLibraries.Select(x => x.FullName));
         }
 
-    #region Utility methods
 
-        private static string DoPackageFileOperationsAndZip(string rootPkgDir, CustomNodeManager customNodeManager, PackageUploadRequestBody header, Package pkg, List<string> files, PackageUploadHandle uploadHandle, bool isTestMode)
+        internal PackageUpload NewPackageUpload(Package package, string packagesDirectory, IEnumerable<string> files, PackageUploadHandle handle)
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageUploadBuilder.cs]: `DoPackageFileOperationsAndZip` is moved to `PackageDirectoryBuilder` as `BuildDirectory`.  File compression, which is necessary for the upload, is moved here.

File: src/DynamoCore/PackageManager/PackageUploadBuilder.cs [hunk 6]
```diff
@@ -1,205 +1,93 @@
...
+            // the file is stored in a tempt directory, we allow it to get cleaned up by the user later
+            if (info.Length > MaximumPackageSize) throw new Exception(Properties.Resources.PackageTooLarge);
 
-        public static string NormalizePath(string path)
-        {
-            return Path.GetFullPath(new Uri(path).LocalPath)
-                       .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
-                       .ToUpperInvariant();
+            return info;
         }
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageUploadBuilder.cs]: Rather than looking at the diff here, it might be more useful to look at the file, which is now rather small.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 7]
```diff
@@ -64,13 +66,15 @@ public bool HasAuthProvider
 
         #endregion
 
-        internal PackageManagerClient(IGregClient client, string rootPackageDirectory, CustomNodeManager customNodeManager)
+        internal PackageManagerClient(IGregClient client, PackageUploadBuilder builder)
         {
-            this.rootPackageDirectory = rootPackageDirectory;
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Check out the new code in DynamoModel.  This is a much cleaner piece of dependency injection.

File: src/DynamoCore/PackageManager/PackageDirectoryBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        /// </summary>
+        /// <param name="package">The package to be formed</param>
+        /// <param name="targetDirectory">The directory where the package should end up.  It is
+        ///     assumed that this directory is empty.  If it doesn't exist,
+        ///     it will be created.
+        /// </param>
+        /// <param name="files">The collection of files to be moved</param>
+        /// <returns></returns>
+        internal IDirectoryInfo BuildDirectory(Package package, string packagesDirectory, IEnumerable<string> files)
+        {
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageDirectoryBuilder.cs]: All of the code here is basically copy pasted from `PackageUploadBuilder.cs` and thoroughly mocked.

File: src/DynamoCore/DynamoCore.csproj [hunk 9]
```diff
@@ -123,7 +123,17 @@ limitations under the License.
...
+    <Compile Include="PackageManager\Interfaces\IFileInfo.cs" />
+    <Compile Include="PackageManager\Interfaces\IFileSystem.cs" />
+    <Compile Include="PackageManager\MutatingFileCompressor.cs" />
+    <Compile Include="PackageManager\MutatingFileSystem.cs" />
+    <Compile Include="PackageManager\MutatingPathRemapper.cs" />
     <Compile Include="PackageManager\PackageManagerResult.cs" />
+    <Compile Include="PackageManager\PackageUploadBuilder.cs" />
+    <Compile Include="PackageManager\TrueDirectoryInfo.cs" />
+    <Compile Include="PackageManager\TrueFileInfo.cs" />
     <Compile Include="Search\Interfaces\ISearchCategory.cs" />
```
Comment (pboyer) [inline src/DynamoCore/DynamoCore.csproj]: Almost all of these new files are required for mocking purposes and should be very straightforward to inspect.

File: src/DynamoCore/PackageManager/MutatingFileCompressor.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Dynamo.PackageManager
+{
+    /// <summary>
+    ///     An IFileCompressor that actually attempts to compress a directory
```
Comment (ikeough) [inline src/DynamoCore/PackageManager/MutatingFileCompressor.cs]: So it's you who has been putting these spaces in their doc comments? What's that all about?

Comment (pboyer) [inline src/DynamoCore/PackageManager/MutatingFileCompressor.cs]: Actually, @Steell started it.  I think it's more readable personally.

File: test/DynamoCoreTests/PackageManager/PackageLoaderTests.cs [hunk 11]
```diff
@@ -7,7 +7,7 @@
 
 namespace Dynamo.Tests
 {
-    class PackageLoaderTests : DynamoViewModelUnitTest
+    class PackageLoaderTests : DynamoModelTestBase
```
Comment (ikeough) [inline test/DynamoCoreTests/PackageManager/PackageLoaderTests.cs]: Hooray.

File: src/DynamoCore/PackageManager/TrueDirectoryInfo.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+﻿using Dynamo.PackageManager.Interfaces;
+
+namespace Dynamo.PackageManager
+{
+    /// <summary>
+    ///     An IDirectoryInfo representing a real DirectoryInfo object
+    /// </summary>
+    internal class TrueDirectoryInfo : IDirectoryInfo
```
Comment (ikeough) [inline src/DynamoCore/PackageManager/TrueDirectoryInfo.cs]: Then shouldn't it be a `RealDirectoryInfo`?

**Comment by ikeough** [comment]
@pboyer Many minor comments mostly around naming.

File: src/DynamoCore/PackageManager/TrueDirectoryInfo.cs (hunk 12 above)
Comment (pboyer) [inline src/DynamoCore/PackageManager/TrueDirectoryInfo.cs]: Yeah, good point.

**Comment by pboyer** [comment]
@ikeough I've your addressed your comments and should be ready for a final look.

**Comment by pboyer** [comment]
ping @ikeough

**Comment by ikeough** [comment]
lgtm.
